### ISSUE TITLE
Address ranges for some entries in location lists from getLocationLists() look incorrect

### ISSUE THREAD

**Issue by wcohen** [dyninst/dyninst#893]
Hi,
I have been working on some code to look at the locations and see if the information lines up with what is actually in the binary.  Some of the address ranges returned by getLocationLists() do not look right.  This can be seen in one of the programs locrangechecks.C is in:

https://github.com/wcohen/quality_info/tree/master/dyninsttools

One can try it out on a simple heapsort program in a src rpm using the installed dyninst-10.1.0-5.fc32.x86_64 on Fedora 32.  The heapsort program can be downloaded from:

https://sourceware.org/systemtap/wiki/AddingUserSpaceProbingToApps?action=AttachFile&do=view&target=heapsort-0.5-2.fc32.src.rpm

rpm -Uvh heapsort-0.5-2.fc32.src.rpm
sudo dnf builddep  ./heapsort-0.5-2.fc32.src.rpm
rpm -Uvh  ./heapsort-0.5-2.fc32.src.rpm
cd ~/rpmbuild/SPECS
rpmbuild -bc heapsort.spec

Build locrangechecks program with a "make" in its directory, then try:

./locrangechecks ~/rpmbuild/BUILD/heapsort-0.5/heapsort

In the output see something like the following for main.  The on_line variable is for the entire range of the main function [4011c0, 40124a], but the argc and argv ranges do not look correct There is no code in the range [802384, 80239f][:

# main [4011c0,40124a]
argc [802384, 80239f] x86_64::rdi
argc [802384, 80239f] x86_64::rdi
argv [802384, 80239f] x86_64::rsi
argv [802384, 80239f] x86_64::rsi
on_line [4011c0, 40124a] 8(x86_64::rsp)

Other functions in the same binariy have similar odd ranges.  It looks there might be some being constant being added to the ranges.
Is there something that the locrangechecks is doing wrong?
Is there some other issue in dyninst causing the problem?

There is a test in the dyninst testsuite that uses getLocationLists().  However, I have not tried it out.

**Comment by hainest** [comment]
@wcohen I'm not able to reproduce this using the current HEAD of master and gcc-9.3.0. I see

	$ ./locrangecheck heapsort-0.5/heapsort
		unable to find targ1080
		# _init [1000,10e0]
		unable to find __printf_chk
		unable to find _Znam
		unable to find __stack_chk_fail
		unable to find _ZdaPv
		unable to find __isoc99_fscanf
		# main [10e0,1136]
		argc [10e0, 10fc] x86_64::rdi
		argv [10e0, 10fc] x86_64::rsi
		on_line [10e0, 10e8] -40(x86_64::rsp)
		on_line [10e8, 1130] 0(x86_64::rsp)
		on_line [1130, 1131] -40(x86_64::rsp)
		on_line [1131, 1136] 0(x86_64::rsp)
		...

The heapsort executable is PIE (although the elf flag isn't set...). Recompiling with '-fno-PIE -no-pie' saw no change.

	$ readelf -h heapsort-0.5/heapsort
	ELF Header:
		Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
		Class:                             ELF64
		Data:                              2's complement, little endian
		Version:                           1 (current)
		OS/ABI:                            UNIX - System V
		ABI Version:                       0
		Type:                              DYN (Shared object file)
		Machine:                           Advanced Micro Devices X86-64
		Version:                           0x1
		Entry point address:               0x1140
		Start of program headers:          64 (bytes into file)
		Start of section headers:          24528 (bytes into file)
		Flags:                             0x0
		Size of this header:               64 (bytes)
		Size of program headers:           56 (bytes)
		Number of program headers:         13
		Size of section headers:           64 (bytes)
		Number of section headers:         38
		Section header string table index: 37

**Comment by wcohen** [comment]
Fedora 32 has gcc-10.2.1.  I just ran things on Fedora 31 with gcc-9.3.1-2.fc31.x86_64 and dyninst-10.1.0-4.fc31.x86_64.  The results look better, so it might be something related to gcc 10.  The `rpm --eval %optflags` used to build the heapsort are identical for f31 and f32.

Output from the f31 version:
```
 ./locrangechecks ~/rpmbuild/BUILD/heapsort-0.5/heapsort
# _init [1000,10f0]
unable to find targ1090
unable to find targ10a0
unable to find targ10b0
unable to find targ10c0
unable to find targ10d0
unable to find targ10e0
# main [10f0,1146]
argc [10f0, 110c] x86_64::rdi
argv [10f0, 110c] x86_64::rsi
on_line [10f0, 10f8] -40(x86_64::rsp)
on_line [10f8, 1140] 0(x86_64::rsp)
on_line [1140, 1141] -40(x86_64::rsp)
on_line [1141, 1146] 0(x86_64::rsp)
# _start [1150,117f]
# deregister_tm_clones [1180,11b0]
# register_tm_clones [11b0,11f0]
# __do_global_dtors_aux [11f0,1230]
# frame_dummy [1230,1240]
# input_data [1240,134a]
data [1263, 1286] (0)
data [1286, 1288] x86_64::rax
data [1288, 12aa] x86_64::r15
data [12aa, 12c9] x86_64::r15
data [12c9, 130d] x86_64::r15
data [130d, 1316] 8(x86_64::rsp)
data [131c, 131e] x86_64::r15
data [131e, 1331] x86_64::r15
data [131e, 1331] x86_64::r15
capacity [1286, 12aa] (2)
capacity [12aa, 12dc] x86_64::rbp
capacity [130d, 1316] x86_64::rbp
capacity [131c, 131e] (2)
capacity [131c, 131e] (2)
value [1240, 1246] -68(x86_64::rsp)
value [1246, 1248] -60(x86_64::rsp)
value [1248, 124a] -52(x86_64::rsp)
value [124a, 124c] -44(x86_64::rsp)
value [124c, 124d] -36(x86_64::rsp)
value [124d, 124e] -28(x86_64::rsp)
value [124e, 1252] -20(x86_64::rsp)
value [1252, 1337] 20(x86_64::rsp)
value [1337, 133b] -20(x86_64::rsp)
value [133b, 133c] -28(x86_64::rsp)
value [133c, 133e] -36(x86_64::rsp)
value [133e, 1340] -44(x86_64::rsp)
value [1340, 1342] -52(x86_64::rsp)
value [1342, 1344] -60(x86_64::rsp)
value [1344, 1345] -68(x86_64::rsp)
value [1345, 134a] 20(x86_64::rsp)
success [128d, 12aa] x86_64::rax
success [12d0, 12d5] x86_64::rax
success [12d5, 1316] (1)
success [131c, 1331] x86_64::rax
success [131c, 1331] x86_64::rax
i [12e8, 12ea] (0)
i [12e8, 12ea] (0)
# output_data [1350,1390]
i [1359, 136f] (0)
i [1359, 136f] (0)
# insert_heap [1390,140e]
start [1390, 13b3] x86_64::rdx
start [13b3, 13bb] x86_64::r8
start [13bb, 1400] x86_64::rdx
start [1405, 140e] x86_64::rdx
start [1405, 140e] x86_64::rdx
m [1399, 13a5] x86_64::r9
m [13a5, 13b3] x86_64::r8
m [13b3, 13d1] x86_64::r9
m [13d1, 1400] x86_64::r8
m [1405, 140e] x86_64::r9
m [1405, 140e] x86_64::r9
# build_heap [1410,1453]
i [141c, 1429] x86_64::rcx
i [1429, 143e] x86_64::rbx
i [143e, 1449] x86_64::rdx
i [1449, 144a] 1(x86_64::rbx)
i [144a, 144f] x86_64::rbx
i [144a, 144f] x86_64::rbx
# heap_sort [1460,14a3]
i [146a, 146f] 1(x86_64::rdi)
i [146f, 1487] x86_64::rcx
i [1487, 149c] 1(x86_64::rcx)
i [149c, 14a3] x86_64::rcx
i [149c, 14a3] x86_64::rcx
# __libc_csu_init [14b0,1515]
# __libc_csu_fini [1520,1525]
# _fini [1528,1528]


```

**Comment by hainest** [comment]
Well, that's odd. I just ran HEAD of master with gcc-10.1.0 and got the same answer as using gcc-9.3.0. I also tested Dyninst-10.1 using gcc-10.1.0 and got the same answer. PIE made no difference in all cases. I am a bit stumped.

**Comment by hainest** [comment]
Confirmed. I see the same issue when building heapsort with clang-10.

**Comment by wcohen** [comment]
I did some more poking around to see what is going on with the problematic clang generated code generated dumps of the dwarf information for the clang and gcc versions with:

$ dwarfdump ./heapsort > heapsort.dump

It looks like clang is encoding the location lists as offsets from the beginning of the function and dyninst is adding the starting address twice rather than just once:

< 1><0x000007be>    DW_TAG_subprogram
                      DW_AT_low_pc                0x004011c0
                      DW_AT_high_pc               <offset-from-lowpc>138
                      DW_AT_frame_base            len 0x0001: 57: DW_OP_reg7
                      DW_AT_GNU_all_call_sites    yes(1)
                      DW_AT_name                  main
                      DW_AT_decl_file             0x00000001 /home/wcohen/rpmbuild/BUILD/heapsort-0.5/heapsort.cxx
                      DW_AT_decl_line             0x0000001c
                      DW_AT_type                  <0x0000005b>
                      DW_AT_external              yes(1)
< 2><0x000007d7>      DW_TAG_formal_parameter
                        DW_AT_location              <loclist at offset 0x00000000 with 1 entries follows>
			[ 0]< offset pair low-off : 0x00000004 addr  0x004011c4 high-off  0x0000001f addr 0x004011df>DW_OP_reg5
                        DW_AT_name                  argc
                        DW_AT_decl_file             0x00000001 /home/wcohen/rpmbuild/BUILD/heapsort-0.5/heapsort.cxx
                        DW_AT_decl_line             0x0000001c
                        DW_AT_type                  <0x0000005b>
< 2><0x000007e6>      DW_TAG_formal_parameter
                        DW_AT_location              <loclist at offset 0x00000023 with 1 entries follows>
			[ 0]< offset pair low-off : 0x00000004 addr  0x004011c4 high-off  0x0000001f addr 0x004011df>DW_OP_reg4
                        DW_AT_name                  argv
                        DW_AT_decl_file             0x00000001 /home/wcohen/rpmbuild/BUILD/heapsort-0.5/heapsort.cxx
                        DW_AT_decl_line             0x0000001c
                        DW_AT_type                  <0x000004c7>

When looking at the same part of the dwarf for the gcc10 version it looks like the offsets are from the beginning of the file:

< 1><0x00000df5>    DW_TAG_subprogram
                      DW_AT_external              yes(1)
                      DW_AT_name                  main
                      DW_AT_decl_file             0x00000001 /home/wcohen/rpmbuild/BUILD/heapsort-0.5/heapsort.cxx
                      DW_AT_decl_line             0x0000001c
                      DW_AT_decl_column           0x00000005
                      DW_AT_type                  <0x00000061>
                      DW_AT_low_pc                0x000010f0
                      DW_AT_high_pc               <offset-from-lowpc>86
                      DW_AT_frame_base            len 0x0001: 9c: DW_OP_call_frame_cfa
                      DW_AT_GNU_all_call_sites    yes(1)
                      DW_AT_sibling               <0x00000e8e>
< 2><0x00000e17>      DW_TAG_formal_parameter
                        DW_AT_name                  argc
                        DW_AT_decl_file             0x00000001 /home/wcohen/rpmbuild/BUILD/heapsort-0.5/heapsort.cxx
                        DW_AT_decl_line             0x0000001c
                        DW_AT_decl_column           0x0000000e
                        DW_AT_type                  <0x00000061>
                        DW_AT_location              <loclist at offset 0x000005f5 with 2 entries follows>
			[ 0]< offset pair low-off : 0x000010f0 addr  0x000010f0 high-off  0x0000110c addr 0x0000110c>DW_OP_reg5
			[ 1]< offset pair low-off : 0x0000110c addr  0x0000110c high-off  0x00001146 addr 0x00001146>DW_OP_GNU_entry_value 0x00000001 contents 0x55 DW_OP_stack_value
                        <Unknown AT value 0x2137>   0x000005f1
< 2><0x00000e2b>      DW_TAG_formal_parameter
                        DW_AT_name                  argv
                        DW_AT_decl_file             0x00000001 /home/wcohen/rpmbuild/BUILD/heapsort-0.5/heapsort.cxx
                        DW_AT_decl_line             0x0000001c
                        DW_AT_decl_column           0x0000001a
                        DW_AT_type                  <0x000007dc>
                        DW_AT_location              <loclist at offset 0x00000632 with 2 entries follows>
			[ 0]< offset pair low-off : 0x000010f0 addr  0x000010f0 high-off  0x0000110c addr 0x0000110c>DW_OP_reg4
			[ 1]< offset pair low-off : 0x0000110c addr  0x0000110c high-off  0x00001146 addr 0x00001146>DW_OP_GNU_entry_value 0x00000001 contents 0x54 DW_OP_stack_value
                        <Unknown AT value 0x2137>   0x0000062e

**Comment by wcohen** [comment]
Looking at the output of " eu-readelf --debug-dump=loc heapsort" for clang compiled version see the base being set to the start of the main:

DWARF section [35] '.debug_loc' at offset 0x5796:

 CU [     b] base: 0x00000000004011c0 <main>
 [     0] range 4, 1f
          0x00000000004011c4 <main+0x4>..
          0x00000000004011de <main+0x1e>
           [ 0] reg5
 [    23] range 4, 1f
          0x00000000004011c4 <main+0x4>..
          0x00000000004011de <main+0x1e>
           [ 0] reg4
 [    46] range 37, 42
          0x00000000004011f7 <main+0x37>..
          0x0000000000401201 <main+0x41>

For gcc see that base offset being 0.  0+0 added to the range values would still give the correct result for the ranges.

Output of "eu-readelf --debug-dump=loc heapsort" for the gcc compiled one:

DWARF section [37] '.debug_loc' at offset 0x5fc2:

 CU [     b] base: +000000000000000000
 [     0] view pair 0, 0
          view pair 0, 0
          view pair 0, 0
 [     6] range 1420, 143e
          +0x0000000000001420 <_Z10build_heapP9list_type>..
          +0x000000000000143d <_Z10build_heapP9list_type+0x1d>
           [ 0] reg5
          range 143e, 1464
          +0x000000000000143e <_Z10build_heapP9list_type+0x1e>..
          +0x0000000000001463 <_Z10build_heapP9list_type+0x43>
           [ 0] reg6
          range 1464, 146a
          +0x0000000000001464 <_Z10build_heapP9list_type+0x44>..
          +0x0000000000001469 <_Z10build_heapP9list_type+0x49>
           [ 0] GNU_entry_value:
                [ 0] reg5
           [ 3] stack_value

**Comment by wcohen** [comment]
Turned on diagnostic printing in dyninst and ran with one of of the other dyninsttool programs, readonly:

$ COMMON_DEBUG_DWARF=true ./readonly -d ~/rpmbuild/BUILD/heapsort-0.5_clang10/heapsort >& dyninst_debug.log

The output shows that the range has already adjusted and then is adjusted a second time.  Below are a couple examples from the heapsort program compiled with clang:

[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:1806] (0x7d7) Decoding loclist location
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:2435] (0x7d7) Decoding entry 1 of 1 over range 0x4011c4 - 0x4011df, mod 0x4011c0 - 0x5f6, base 0x4011c0
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/dwarf/src/dwarfExprParser.C:83] Entry to decodeDwarfExpression
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/dwarf/src/dwarfExprParser.C:95] 	Atom 1 of 1: val 0x55
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/dwarf/src/dwarfExprParser.C:107] 		 Pushing reg x86_64::rdi
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/dwarf/src/dwarfResult.C:49] 		Push x86_64::rdi
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:2483] (0x7d7) Using lexical range, shifted by module low
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:2487] (0x7d7) valid over range 0x802384 to 0x80239f
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:1445] (0x7d7) Checking for abstract origin
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:1512] (0x7d7) Found name argc

[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:1806] (0x7e6) Decoding loclist location
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:2435] (0x7e6) Decoding entry 1 of 1 over range 0x4011c4 - 0x4011df, mod 0x4011c0 - 0x5f6, base 0x4011c0
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/dwarf/src/dwarfExprParser.C:83] Entry to decodeDwarfExpression
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/dwarf/src/dwarfExprParser.C:95] 	Atom 1 of 1: val 0x54
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/dwarf/src/dwarfExprParser.C:107] 		 Pushing reg x86_64::rsi
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/dwarf/src/dwarfResult.C:49] 		Push x86_64::rsi
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:2483] (0x7e6) Using lexical range, shifted by module low
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:2487] (0x7e6) valid over range 0x802384 to 0x80239f
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:1445] (0x7e6) Checking for abstract origin
[/home/wcohen/rpmbuild/BUILD/dyninst-10.2.1/dyninst-10.2.1/symtabAPI/src/dwarfWalker.C:1512] (0x7e6) Found name argv

Something is not right with the logic in DwarfWalker::decodeLocationListForStaticOffsetOrAddress()

**Comment by hainest** [comment]
@wcohen Thanks for chasing this down! I'll try to make this a priority after the 11.0 release. There is a lot going on in Dyninst right now, so it might be a while before I can get to it.

**Comment by wcohen** [comment]
@hainest Thanks.  I can see see where things are going wrong in  DwarfWalker::decodeLocationListForStaticOffsetOrAddress where the base is always being added in the section of code below.  For the gcc generated code the base is 0, so it turns out things are okay.  However, for clang the ranges have already been adjusted to include the base as seen in the earlier comment and then it gets added again.  I don't know if just removing the "+ base" in the statements below is going to cause some other problem.
`
            dwarf_printf("(0x%lx) Using lexical range, shifted by module low\n", id());
            loc.lowPC = location->ld_lopc + base;
            loc.hiPC = location->ld_hipc + base;

            dwarf_printf("(0x%lx) valid over range 0x%lx to 0x%lx\n",
                    id(), loc.lowPC, loc.hiPC);
            locs.push_back(loc);
`
That came from  the following commit. Maybe this is a misunderstanding how libdw returns range information.

commit 18278f6b4816d21f4dff56739bfe6efe5c3c0c16
Author: Sasha Nicolas <sasha@cs.wisc.edu>
Date:   Thu Feb 23 19:12:13 2017 -0600

    New changes to adapt code to libdw.

**Comment by hainest** [comment]
@sashanicolas Could you take a look at this? From WIll's comments, it seems the fix should be fairly straightforward. We won't be able to test this thoroughly until I get more of the clang stuff working in Dyninst.

**Comment by hainest** [comment]
Fixed by #1297
